javascript/src/refactor.js

```diff
@@ -0,0 +1,648 @@
+const winston = require('winston'); // logger
+const colors = require('colors');
+
+class RefactorEngine {
+    constructor(ast, 
+                properties, 
+                behaviour, 
+                flagname, 
+                max_cleanup_steps, 
+                print_to_console = false, 
+                keep_comments = false,
+                filename = null) {
+        this.ast = ast;
+        this.properties = properties;
+        this.flagname = flagname;
+        this.max_cleanup_steps = max_cleanup_steps;
+        this.print_to_console = print_to_console;
+        this.keep_comments = keep_comments;
+        this.behaviour = behaviour;
+        this.changed = false;
+        this.filename = filename;
+    }
+        
+    trueLiteral() { 
+        const vanillaLiteral = { // A Boolean Literal is parsed like this by recast
+            type: "Literal", 
+            value: true, 
+            raw: 'true'
+        };
+    
+        // This additional property marks that this literal was introduced by Piranha 
+        vanillaLiteral.createdByPiranha = true;
+    
+        return vanillaLiteral;
+    }
+    
+    falseLiteral() { 
+        const vanillaLiteral = { // A Boolean Literal is parsed like this by recast
+            type: "Literal", 
+            value: false, 
+            raw: 'false'
+        };
+    
+        // This additional property marks that this literal was introduced by Piranha 
+        vanillaLiteral.createdByPiranha = true;
+    
+        return vanillaLiteral;
+    }
+
+    getMethodHashMap(properties) { 
+        var methodHashMap = new Map();
+    
+        properties["methodProperties"].forEach((item) => {
+            methodHashMap.set(item["methodName"], item);
+        });
+    
+        return methodHashMap;
+    }
+
+    // Verify this is a literal introduced by Piranha. 
+    // We only refactor code associated with literals having this property and leave original code untouched
+    isPiranhaLiteral(node) {
+        return node.type === "Literal" && node.createdByPiranha !== undefined;
+    }
+
+    reduceLogicalExpression(literal, expression, operator) {
+        if (operator === "&&") {
+            if (literal.value === true) {
+                return expression;
+            }
+                
+            if (literal.value === false) {
+                return this.falseLiteral();
+            }
+        } else if (operator === "||") {
+            if (literal.value === true) {
+                return this.trueLiteral();
+            }
+                
+            if (literal.value === false) {
+                return expression;
+            }
+        }
+    }
+
+    // Check if a function with return statement at the end of function body is redundant, i.e returns piranhaLiteral
+    checkAndAddRedundantFunction(func, name, redundantFunctions) {
+        if (func.body === null || func.body.body == null) {
```

Comment:
Curious here, what does `func.body.body` mean here and in the rest? Why not just `func.body`?

Reply:
This is a quirk of nomenclature. The `body` of a `FunctionDeclaration` refers to the contents of the function. When a function is enclosed within braces `{ ... }`, its `body` is a `BlockStatement` node. The `BlockStatement` itself contains a `body` attribute which is an array of nodes corresponding to each line of code. Therefore, `func.body.body` refers to this array.